Support "rules" YAML input for ast-grep

jayeheffernan opened the pull request:
This PR introduces `astgrep-rules` as a 3rd engine. See #363 for discussion and motivations.

### New engine, new inputs

When this engine is selected, enter astgrep rules YAML into the main "Rules" input. The input will be highlighted as YAML. The "Search" field is replaced by "Rules", and "Replace" is hidden because that can be handled by `fix` fields within the YAML.

### How to use

Open grug-far, `<localleader>e` to switch from `ripgrep` to `astgrep`, then again to switch to `astgrep-rules`. See [astgrep rules docs](https://ast-grep.github.io/guide/rule-config.html) for inspiration on how to search.

### Default input value

There is some smarts to set the default input value, when switching to `astgrep-rules`. If you've been working on a pattern in the `astgrep` engine, you may want to keep that in your rules YAML and extend it, so when switching from `Search: $MYPATTERN` with `astgrep` to `astgrep-rules` the `Rules:` input will contain `pattern: $MYPATTERN`. Similarly, `fix: ` is filled from `Replacement: `, if it exists.

The YAML requires a `language` field too. Setting a default value here is a little more complicated. This can be filled as `filetype` from the file that was being edited - but it is possible to do better... The astgrep CLI supports a project-wide (or global) config file, including a field `languageGlobs` that can have entries like `html: ['*.vue', '*.svelte', '*.astro']`. This says e.g. "use HTML as the Treesitter language when parsing a `.vue` file". If the user has that in their config, and is editing a Vue file when they trigger grug-far, it would be better to set `language: html`, rather than e.g. `language: vue`. In my local config, I'll be defining JS, TS, JSX, and TSX to all be parsed as TSX (the superset language) - so that I don't have to rewrite rules 4 times to match properly across all these similar file types in my project. It would be ideal if grug-far can read and interpret the globs from the YAML file, but I don't think it's worth it to do YAML parsing here? We could ignore the whole issue and not map based on globs at all. I've gone a middle-road: the user can configure `grug-far` with a Lua table of `languageGlobs`, and we map based on that to set the default `language`. If no glob matches, or none are configured, then `filetype` is the fallback.

#### Example workflow

Say you're editing a `.ts` Typescript file, and you have configured `grug-far`'s `astgrep-rules` engine with `languageGlobs = { tsx = { "*.ts", "*.tsx" } }`. You open `grug-far` You switch to the `astgrep` engine and type `$A < $B` and `$B > $A` as a replacement. Now you switch engine to `astgrep-rules`. The search and replacement fields disappear, and your new "Rules" input would contains this:

```yaml
id: my_rule_1
language: tsx
rule:
  pattern: $A < $B
fix: $B > $A
```

This should be equivalent to what you were doing, but you can now extend that YAML with any additional features of rules.

### Other notes

- The default language detection logic is all run _when switching to the astgrep-rules engine_, and is based on looking at the _current_ buffer in the `prevWin` window. The buffer in that window could have changed since `grug-far` was opened. Do we want to track a `prevFilename` for this, so that we remember exactly what file was being edited when `grug-far` was opened?
- I don't have `stylua` and `selene` running, so I'm counting on the Github actions to pick up any issues there. Hopefully that works out
- I've not updated any tests or docs yet. Figure it makes sense to settle on the implementation first. If you want this done, LMK
- astgrep does not support `--inline-rules` more than once, or together with `--rule`, so these are blacklisted
- I tried logic like `local isRuleMode = inputs.rules ~= nil`, but this didn't work out. I think the problem is, rotating through different engines, `inputs` will then contain both `.rules` and `.search` (whatever the last value was of each). Not sure if that's something you want to address, or happy to leave as-is
- I've not tried to implement any YAML LSP. otter.nvim does this on partial buffers by keeping a separate hidden buffer for each code-block, and redirecting LSP requests between them. Doesn't seem worth it to me to try to implement here. If we really want LSP on the Search/Rules input, I would suggest to explicitly make it a separate window and buffer (and then LSP is trivial), rather than trying to keep all the grug-far inputs and results in the one buffer
- Now that there are 3 engines, the order matters. This was previously based on iterating table keys in a way where order wasn't defined. This was giving me an order of ripgrep, astgrep-rules, astgrep. I think it makes more sense to be ripgrep, astgrep, astgrep-rules. I've defined this order in an entirely new config key. If you'd prefer I assign sort-indicies within each engine, or turn `.engines` into an array (with `.name`) fields, LMK
- Suggestion: run `getDefaultValue()` every time we switch engine, but give that function access to the current value. Or treat "existing value, empty string" the same as `nil`, when choosing whether to run `getDefaultValue()`. That way if the rules input is empty (but has been filled previously), and we edit search again in `astgrep`, then switch to `astgrep-rules` again, we can re-derive the default-value with the new search

### ⚠️ Bugs

#### Truncated YAML input

I don't think I introduced this, but it's hitting me. Something about how the rules input renders seems to cut it off to 4 lines. That's a problem for the default value if there is some existing replacement value, which adds a `fix:` line to the YAML. It can also show up like this:

1. Switch to `astgrep-rules` engine
2. Manually type in YAML, longer than 4 lines
3. Switch engines, 3 times, to come back to the same engine
4. Notice the YAML is truncated

#### "show command" doesn't show

`<localleader>p` notifies that this is toggled on, but it doesn't actually show the command. I haven't lookend into it much, because maybe it's related to the truncated YAML bug. Maybe not.

### Screenshot

Here's a screenshot of it in action, searching this repo. Note that replace with `<leader>r` does work. I've not come up with a realistic example here, but it demos the syntax a little.

<img width="1754" alt="Screenshot 2025-03-02 at 19 00 53" src="https://github.com/user-attachments/assets/e213f64a-59cc-4f16-bdd5-44a3ac1d032b" />

Some example text you might search with (not actually realistic, but it works):

```yaml
id: test
language: lua
rule:
  pattern: $A
  kind: dot_index_expression
  inside:
    kind: if_statement
    stopBy: end
fix: ($A and $A)
```

MagicDuck reviewed (commented):
good start! :)

MagicDuck commented:
> I tried logic like local isRuleMode = inputs.rules ~= nil, but this didn't work out. I think the problem is, rotating through different engines, inputs will then contain both .rules and .search (whatever the last value was of each). Not sure if that's something you want to address, or happy to leave as-is
fixed here: https://github.com/MagicDuck/grug-far.nvim/pull/369

MagicDuck commented:
> The default language detection logic is all run when switching to the astgrep-rules engine, and is based on looking at the current buffer in the prevWin window. The buffer in that window could have changed since grug-far was opened. Do we want to track a prevFilename for this, so that we remember exactly what file was being edited when grug-far was opened?

sure, sounds good!

MagicDuck commented:
> I've not tried to implement any YAML LSP. otter.nvim does this on partial buffers by keeping a separate hidden buffer for each code-block, and redirecting LSP requests between them. Doesn't seem worth it to me to try to implement here. If we really want LSP on the Search/Rules input, I would suggest to explicitly make it a separate window and buffer (and then LSP is trivial), rather than trying to keep all the grug-far inputs and results in the one buffer

Yeah, I don't know that we can do that atm while keeping the current editing model. If neovim starts supporting some clever way to move between buffers at some point or we find some smart way of doing it, we can think about it then.

MagicDuck commented:
> Now that there are 3 engines, the order matters. This was previously based on iterating table keys in a way where order wasn't defined. This was giving me an order of ripgrep, astgrep-rules, astgrep. I think it makes more sense to be ripgrep, astgrep, astgrep-rules. I've defined this order in an entirely new config key. If you'd prefer I assign sort-indicies within each engine, or turn .engines into an array (with .name) fields, LMK

That's totally fine, in fact it's good to have an array where people can define the engines they want to loop through. Like if they never want to swap to astgrep, they can take that out.

MagicDuck commented:
> Suggestion: run getDefaultValue() every time we switch engine, but give that function access to the current value. Or treat "existing value, empty string" the same as nil, when choosing whether to run getDefaultValue(). That way if the rules input is empty (but has been filled previously), and we edit search again in astgrep, then switch to astgrep-rules again, we can re-derive the default-value with the new search

Is this fixed by https://github.com/MagicDuck/grug-far.nvim/pull/369 ? Otherwise, it might be a good idea. Note that `getDefaultValue()` is also called initially if you launch directly with `engine: 'astgrep-rules'` option

MagicDuck commented:
> bugs: Truncated YAML input

I cannot reproduce this anymore after https://github.com/MagicDuck/grug-far.nvim/pull/369
(without your changes, just renamed existing astgrep search input to "rules")

MagicDuck commented on lua/grug-far/engine/astgrep-rules.lua:
```diff
@@ -0,0 +1,142 @@
+      trim = false,
+      getDefaultValue = function(context)
+        local lang = ''
+        if context.prevWin ~= nil then
```
1. would move this logic to get the lang to function 
2. agree that we can track prevFiletype or prevBuffer so that it does not matter if it changes in between

jayeheffernan replied:
done and done

MagicDuck commented on lua/grug-far/engine/astgrep-rules.lua:
```diff
@@ -0,0 +1,142 @@
+
+  isSearchWithReplacement = function(i, options)
+    local inputs = shallow_copy(i)
+    inputs.isRuleMode = true
```
hopefully we can remove all those shallow copys and `isRuleMode` now

MagicDuck commented on lua/grug-far/engine/astgrep/getArgs.lua:
```diff
@@ -14,13 +14,22 @@ local rewriteFlags = {
 ---@return string[]? args, string[]? blacklisted
 local function getArgs(inputs, options, extraArgs, blacklistedFlags, forceReplace)
-  if #inputs.search < (options.minSearchChars or 1) then
+  local isRuleMode = inputs.isRuleMode or false
```
either inputs.search or inputs.rules should be appropriately nil now

MagicDuck commented on lua/grug-far/opts.lua:
```diff
@@ -95,10 +98,36 @@ M.defaultOptions = {
+        --  rules would normally be multi-line, but we don't support multi-line
+        --  placeholders. rules is filled with a default-value though, so it's
+        --  rare to see it empty
+        rules = 'id: my_rule_1',
```
Maybe something like this would work to give more of an idea to the user that it will be multiline:
`rules = 'id: my_rule_1 \\n rule: \\n  pattern: await $A',...`
But likely it won't matter most of the time since we fill with default value.

MagicDuck reviewed (changes requested):
nice work!

jayeheffernan commented on lua/grug-far/opts.lua:
```diff
@@ -95,10 +98,36 @@ M.defaultOptions = {
+      -- like for example if you always want context lines around matches
+      extraArgs = '',
+
+      languageGlobs = {},
```
done

MagicDuck replied:
sorry to be a pain, but could you add an example of what the user would put in `languageGlobs` to the comment? It's not completely clear.

MagicDuck replied:
super!

jayeheffernan commented on lua/grug-far.lua:
```diff
@@ -102,11 +102,13 @@ local contextCount = 0
 ---@field filesFilter string
 ---@field flags string
 ---@field paths string
+---@field isRuleMode? boolean
```
Yep `input.rules ~= nil` is working now, thanks

jayeheffernan commented:
Thanks @MagicDuck ! All that feedback's addressed, and it's mostly working well.

`getDefaultValue()` is now called every time engine is swapped, if defined. In the default config, it's only defined for `astgrep-rules`'s `rules` input. This decides to keep the previous value if there already is a _non-empty_ one. So if something's been typed (or pre-filled) in the input, then it won't be clobbered as you cycle through engines. But if you delete the content of the input and cycle around, then the default-value can be re-inferred from any existing `search`/`replace` inputs. It strikes me that this is a change in API, so could potentially break a user's config. But you _just_ introduced `getDefaultValue()` and the inputs array, right? In which case it should be fine.

### Bug

I am seeing a bug with input-rendering. I can trigger it with this config:

```lua
{
  engine = "astgrep-rules",
  enabledEngines = { "astgrep-rules", "ripgrep", "astgrep" },
}
```

Open `grug-far` (which starts on `astgrep-rules` and fills the default YAML value), then swap engine once, to land on `ripgrep`. That triggers the bug. Then the "Search:" input starts getting larger. I guess it's adding newlines to the value, about 2 per second.

MagicDuck commented:
Checking out the bug.

btw, I think we need to disable colors in the output:

![image](https://github.com/user-attachments/assets/387911f1-66f7-4eee-85e8-0e7dc1b3b557)

MagicDuck commented:
> I am seeing a bug with input-rendering. I can trigger it with this config:

I believe I have resolved it. Rebasing should fix it! 🤞
See: https://github.com/MagicDuck/grug-far.nvim/pull/373/files

MagicDuck commented:
> But you just introduced getDefaultValue() and the inputs array, right? In which case it should be fine.

Yes, that's totally fine

MagicDuck commented on lua/grug-far.lua:
```diff
@@ -206,6 +209,10 @@ end
 local function createWindow(context)
   context.prevWin = vim.api.nvim_get_current_win()
+  local prevBuf = vim.api.nvim_win_get_buf(context.prevWin)
+  context.prevBufName = vim.api.nvim_buf_get_name(prevBuf)
```
these 2 props need to be added to `GrugFarContext` type in this file.

MagicDuck commented on lua/grug-far/engine/astgrep-rules.lua:
```diff
@@ -0,0 +1,149 @@
+end
+
+---@param context GrugFarContext
+local function get_default_astgrep_language(context)
```
nit: need ---@return

jayeheffernan replied:
this is a stylistic thing I guess, as my LSP was already correctly inferring the return type, so I think it was working for type-hinting. Happy to make it explicit though 👍

MagicDuck replied:
that's fine then, just my OCD 😄

MagicDuck commented on lua/grug-far/engine/astgrep/search.lua:
```diff
@@ -19,7 +19,7 @@ function M.getSearchArgs(inputs, options)
     '--json=stream',
   }
-  return getArgs(inputs, options, extraArgs, blacklistedSearchFlags)
+  return getArgs(inputs, options, extraArgs, blacklistedSearchFlags, false)
```
nit: last arg is not necessary since `forceReplace` is optional, right?

MagicDuck commented on lua/grug-far/engine/astgrep/search.lua:
```diff
@@ -135,7 +135,13 @@ function M.search(params)
   end
-  local numSearchChars = #params.inputs.search
+  local isRuleMode = params.inputs.rules ~= nil
+  local numSearchChars
```
shorter is:
```
local numSearchChars = isRuleMode and #params.inputs.rules or #params.inputs.search
```

jayeheffernan replied:
My misunderstanding coming from JS. I thought this would work like JS's `a && b || c` - which if we have `true && 0 || 4` would eval to `4` - where what I really wanted was JS's `a ? b : c`. I think I see now, Lua doesn't consider `0` or `''` false-y, so `true and b or 4` will only evaluate to `4` if `b` is `false` or `nil` - which works for my uses here

MagicDuck replied:
no worries, I also work with typescript on day to day basis, and I always forget what's falsy in lua, haha

MagicDuck commented on lua/grug-far/engine/astgrep/getArgs.lua:
```diff
@@ -14,13 +14,22 @@ local rewriteFlags = {
-  if #inputs.search < (options.minSearchChars or 1) then
+  local isRuleMode = inputs.rules ~= nil
+
+  local searchInputLen = 0
```
nit, shorter is:
```
local numSearchChars = isRuleMode and #params.inputs.rules or #params.inputs.search
```

MagicDuck commented on lua/grug-far/opts.lua:
```diff
@@ -42,6 +42,9 @@ M.defaultOptions = {
   -- deprecated, please use engines.ripgrep.extraArgs
   extraRgArgs = '',
 
+  -- Order to rotate through engines when swapping engines
```
this is not only order, it's also which engines are enabled for usage. Maybe change wording to something like:
```
engines that are enabled for usage
Also dictates order to rotate through when swapping engines
```

jayeheffernan replied:
Updated. Note that this could be argued though, because it is still possible to activate an unlisted engine by passing it directly to `.open()`, or setting it as `engine=` in your config

MagicDuck reviewed (changes requested):
amazing work! It's getting really close now 👍

I think you can start adding tests at this point.
I am fine with placing them in the tests/astgrep dir.
See this file for how to run them:
https://github.com/MagicDuck/grug-far.nvim/blob/main/CONTRIBUTING.md

Also README.md likely needs updating.

jayeheffernan commented on lua/grug-far/engine/astgrep-rules.lua:
```diff
@@ -0,0 +1,149 @@
+
+  -- If the user has configure any globs to map filenames to languages, we can
+  -- use those
+  if context.prevBufName ~= nil then
```
I don't know if there'd be a situation with no filetype, but this is more to cover situations with `languageGlobs` where we can choose a _better_ language to set for `ast-grep`. E.g. I'm editing a JS file, `filetype` is `javascript`, but my preference with `ast-grep` is to always parse JS as `tsx` - checking the buffer name can let us set `language: tsx`, whether or not there was a reasonable filetype

MagicDuck replied:
gotcha!

jayeheffernan commented:
@MagicDuck I've added test files, by copying the existing `astgrep` ones and then modifying from there. The original copy is a commit of its own, so you could see the diff going forward from there if you like. I did skip a few too - related to different replacement interpreters. With `astgrep-rules` there is no replacement input field, it's all handled by `sg` based on the single YAML input, so I don't think there's any sense in trying to apply different interpreters.

I have updated docs in the README now too, and think I've addressed every smaller bit of feedback. Did not try to add any screenshots to the README, as the existing ones are all nice and consistent in your colorscheme, font, etc.

The `getDefaultValue()` change is working well - now the `astgrep-rules` can re-pre-fill its "rules" input from a changed `astgrep` "search" input, if the user has emptied/cleared their "rules" input (otherwise we don't clobber it).

There are two issues I'm unsure how to address. Please advise 🙏 

## Colors

I see what you mean in the screenshot, it does the same for me. Is this something we can control in Lua after receiving the command output though? Or, do we _want_ to try some regex hack to remove the colour sequences? My understanding is that we want the command output that we receive to not contain any colours (already) - but `sg` doesn't seem to respect `--color=never` in the case of invalid arguments (and yet, other invalid arguments don't display this poorly for me 🤨 )

## Tests

I have some tests failing that I cannot seem to understand. Have a look at a573e50, which I've now reverted. It's giving me the impression that maybe there is still some issue with rendering inputs. The test is trying to pass the `--rule` flag and confirm it sees a warning that that's invalid. The test fails, and I can see that the `--rule` flag is not filled, in the screenshot that prints. But if I just reduce the number of lines in the primary search/rule input (that commit I've linked)... then the test passes 😅 Some other tests are failing while waiting for some text, and the screenshot shows valid/expected input, but with the "INVALID_RULES" error message (the one with the color codes). Copying the same input into my own live nvim session, it works fine (no error about invalid rules).

I'm also seeing a few test failures on `main`, so it's hard to tell if there might be an existing issue that's causing some of my failures.

MagicDuck commented on README.md:
````diff
@@ -209,7 +210,7 @@ Note that you can edit the history buffer and save just like any other buffer if
 ```
 <optional comment, e.g. My special search>
-Engine: <astgrep|ripgrep>(|lua)?
+Engine: <astgrep|astgrep-rules|ripgrep>(|lua)?
````
also add `Rules: <text>`

and add a mention below that some of those fields may vary depending on engine.

MagicDuck commented on README.md:
```diff
@@ -225,8 +226,8 @@ History entries are separated by one or more empty lines.
 
-### Seeing the full rg search command
-Sometimes, mostly for debug purposes, it's useful to see the full `rg` command that gets executed on search. You
+### Seeing the full search command
```
thank you! 👏

MagicDuck commented on README.md:
```diff
@@ -236,7 +237,11 @@ The command is shell-escaped, so you can copy and execute it in a shell manually
 
 ### Swapping search engine
-You can swap search engines with the `Swap Engine` action. Currently `ripgrep` (default) and `astgrep` are supported. 
+You can swap search engines with the `Swap Engine` action. Currently `ripgrep` (default), `astgrep`, and `astgrep-rules` are supported. 
```
nice!

MagicDuck commented on lua/grug-far/inputs.lua:
```diff
@@ -19,7 +19,7 @@ local function fillInput(context, buf, name, value, clearOld)
   if inputRow then
     local oldValue = context.state.inputs[name]
-    local oldNumInputLines = #vim.split(oldValue, '\n')
+    local oldNumInputLines = #vim.split(oldValue or '', '\n')
```
hmm, this change might indicate a bug. I think it might be necessary to go in there and get the value from the buffer directly since if there is any inconsistency, it would screw up the fill logic. 
 
I'll try to reproduce this and fix it myself, but if you can let me know the reproduction steps where you saw this.

MagicDuck commented:
@jayeheffernan 
Amazing work!

### Colors
seems like a regressing in `sg`, unless I am doing something stupid, haha. I have created a issue in their github for now. The author is quite responsive in my experience, so we should hopefully have a fix soon! https://github.com/ast-grep/ast-grep/issues/1850


### Potential oldNumInputLines issue

https://github.com/MagicDuck/grug-far.nvim/pull/364#discussion_r1983846395
I'll investigate

### Tests
Sorry I didn't realize how big this was going to get 😓 
Can we:
1. add a new folder under tests/, let's say astgrep-rules, and move your tests there.
2. add a new "job" that runs those tests only. You can duplicate and rename the existing one for astgrep https://github.com/MagicDuck/grug-far.nvim/blob/1a7377ba5c012b68b835dc95fbf218adbd07afa9/.github/workflows/tests.yml#L39 

I'll take a look at what's failing for you and give suggestions. The mini.tests child neovim instance is not always perfect, so you have to trick it sometimes. There are also sometimes edge cases where there's race conditions that I have missed adding "waits" for, which is maybe why you were seeing some failures in `main`.

MagicDuck commented on tests/astgrep/test_rules_goto.lua:
```diff
@@ -0,0 +1,49 @@
+local MiniTest = require('mini.test')
```
I would remove this one. You don't need to re-test goto as the new engine is not doing anything goto specific.

MagicDuck commented on tests/astgrep/test_rules_open.lua:
```diff
@@ -0,0 +1,47 @@
+local MiniTest = require('mini.test')
```
I would remove this one. You don't need to re-test open as the new engine is not doing anything specific to that feature.

MagicDuck commented on tests/astgrep/test_rules_parseResults.lua:
```diff
@@ -0,0 +1,86 @@
+local parseResults = require('grug-far.engine.astgrep.parseResults')
```
should remove, not necessary imo as it's already tested in existing file.

MagicDuck commented on tests/astgrep/test_rules_sync.lua:
```diff
@@ -0,0 +1,96 @@
+local MiniTest = require('mini.test')
```
I would remove this one. You don't need to re-test sync as the new engine is not doing anything specific to that feature.

MagicDuck commented:
@jayeheffernan There are some issues with history fill (similar to how we had issues with swap engine). I'm attempting to fix them and potentially merge to your branch. I would avoid fixing anything in tests for now.

MagicDuck commented:
@jayeheffernan hi there. I have fixed the input rendering issue and the tests. Was a gnarly rendering timing problem. The changes are in [astgrep-rules](https://github.com/MagicDuck/grug-far.nvim/tree/astgrep-rules) branch in this repo. You can see them if you go to this temporary PR: 
https://github.com/MagicDuck/grug-far.nvim/pull/376

I wasn't sure of the best way to contribute back to this PR, but I figure it should be fairly easy to either fetch that branch and reset hard to it or cherry-pick my changes from that branch. Lemme know if you run into any issues.

jayeheffernan commented:
Nice one @MagicDuck ! Sounds like I might've never got to the bottom of that one 😄  Looks like your issue about the colour codes is already addressed over in the ast-grep repo. I've hard-reset this PR to match your tmp PR, then added one small tweak to fix a test that was failing for me.

I am actually still seeing quite a few test failures locally 😂  but they're passing in CI. I think I see what is causing the rest of them. It's around screenshots (or occasionally waiting for matching text, but it's the same content mismatch) that includes the name of the `ast-grep` command. Here's an example:

<details>
<summary>Test Failure</summary>

```txt
make: *** [test] Error 1
nvim --headless --noplugin -u ./scripts/minimal_init.lua -l ./scripts/test_cli.lua
Total number of cases: 16
Total number of groups: 1

tests/astgrep-rules/test_search.lua: ooxoxooooooooooo

Fails (2) and Notes (0)
FAIL in tests/astgrep-rules/test_search.lua | reports error from sg:
  Failed expectation for screenshot equality to reference at "tests/screenshots/tests-astgrep-rules-test_search.lua---reports-error-from-sg".
  Different `text` cell at line 22 column 8. Reference: "s". Observed: "a".
  Reference:
  --|---------|---------|---------|---------|---------|---------|---------|---------|
  01| Actions / Help g?    Replace ,r    Sync All ,s    Sync Line ,l ...
  02|
  03|  Rules:
  04|id: grug_test
  05|language: typescript
  06|rule:
  07|  pattern: grug
  08|
  09|  Files Filter:
  10|e.g. *.lua   *.{css,js}   **/docs/*.md   (specify one per line, filters vi ...
  11| 󰮚 Flags:
  12|--strictness
  13|  Paths:
  14|e.g. /foo/bar   ../   ./hello\ world/   ./src/foo.lua   ~/.config
  15|
  16| STATUS_ERROR ⟪ astgrep-rules ⟫
  17|
  18|error: unexpected argument '--strictness' found
  19|
  20|  tip: a similar argument exists: '--stdin'
  21|
  22|Usage: sg scan <PATHS|--follow|--no-ignore <FILE_TYPE>|--stdin|--globs <GLOBS@@@
  23|Grug FAR - 1: id: grug_test^@language: typesc...              2,14           Top
  24|-- INSERT --

  --|---------|---------|---------|---------|---------|---------|---------|---------|
  01|00000000000000000112220000000000112220000000000011222000000000000112222222222222
  02|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  03|11111111122222222222222222222222222222222222222222222222222222222222222222222222
  04|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  05|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  06|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  07|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  08|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  09|11111111111111112222222222222222222222222222222222222222222222222222222222222222
  10|33333333333333333333333333333333333333333333333333333333333333333333333333333322
  11|11111111122222222222222222222222222222222222222222222222222222222222222222222222
  12|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  13|11111111122222222222222222222222222222222222222222222222222222222222222222222222
  14|33333333333333333333333333333333333333333333333333333333333333333222222222222222
  15|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  16|33333333333333333333333333333333333333333333333333333333333333333333333333333333
  17|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  18|44444444444444444444444444444444444444444444444222222222222222222222222222222222
  19|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  20|44444444444444444444444444444444444444444442222222222222222222222222222222222222
  21|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  22|44444444444444444444444444444444444444444444444444444444444444444444444444444555
  23|66666666666666666666666666666666666666666666666666666666666666666666666666666666
  24|00000000000077777777777777777777777777777777777777777777777777777777777777777777

  Observed:
  --|---------|---------|---------|---------|---------|---------|---------|---------|
  01| Actions / Help g?    Replace ,r    Sync All ,s    Sync Line ,l ...
  02|
  03|  Rules:
  04|id: grug_test
  05|language: typescript
  06|rule:
  07|  pattern: grug
  08|
  09|  Files Filter:
  10|e.g. *.lua   *.{css,js}   **/docs/*.md   (specify one per line, filters vi ...
  11| 󰮚 Flags:
  12|--strictness
  13|  Paths:
  14|e.g. /foo/bar   ../   ./hello\ world/   ./src/foo.lua   ~/.config
  15|
  16| STATUS_ERROR ⟪ astgrep-rules ⟫
  17|
  18|error: unexpected argument '--strictness' found
  19|
  20|  tip: a similar argument exists: '--stdin'
  21|
  22|Usage: ast-grep scan <PATHS|--follow|--no-ignore <FILE_TYPE>|--stdin|--globs @@@
  23|Grug FAR - 1: id: grug_test^@language: typesc...              2,14           Top
  24|-- INSERT --

  --|---------|---------|---------|---------|---------|---------|---------|---------|
  01|00000000000000000112220000000000112220000000000011222000000000000112222222222222
  02|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  03|11111111122222222222222222222222222222222222222222222222222222222222222222222222
  04|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  05|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  06|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  07|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  08|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  09|11111111111111112222222222222222222222222222222222222222222222222222222222222222
  10|33333333333333333333333333333333333333333333333333333333333333333333333333333322
  11|11111111122222222222222222222222222222222222222222222222222222222222222222222222
  12|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  13|11111111122222222222222222222222222222222222222222222222222222222222222222222222
  14|33333333333333333333333333333333333333333333333333333333333333333222222222222222
  15|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  16|33333333333333333333333333333333333333333333333333333333333333333333333333333333
  17|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  18|44444444444444444444444444444444444444444444444222222222222222222222222222222222
  19|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  20|44444444444444444444444444444444444444444442222222222222222222222222222222222222
  21|22222222222222222222222222222222222222222222222222222222222222222222222222222222
  22|44444444444444444444444444444444444444444444444444444444444444444444444444444555
  23|66666666666666666666666666666666666666666666666666666666666666666666666666666666
  24|00000000000077777777777777777777777777777777777777777777777777777777777777777777
  Traceback:
    ...re/nvim/lazy/grug-far.nvim/lua/grug-far/test/helpers.lua:261
    tests/astgrep-rules/test_search.lua:106
```

</details>

It seems the test screenshots are looking for `Usage: sg scan`, whereas on my system I see `Usage: ast-grep scan`. I'm guessing we're just on different versions, and [this ast-grep update](https://github.com/ast-grep/ast-grep/commit/15295de3f48aa39bee7c2af642fceb7742d9c156) changed it. My current version:

```sh
sg --version
# ast-grep 0.35.0
```

Looks like the Github actions are on 0.31.

LMK if you want:

- any more code changes here
- any clean-up on the commit history. Otherwise I'm assuming just a squash-and-merge into one commit to deploy
- if all tests are passing for you - maybe you want to upgrade ast-grep version locally and then update the screenshots? I don't want to do it from here because I'm not starting from a "clean slate" with all tests passing. But if you update the screenshots, I can confirm they all then pass for me

Thanks for all your help getting this sorted! Seems like it's getting close 🥳

HerringtonDarkholme commented:
Hi ast-grep author here. Thanks for brining it to grug-far!

> it seems the test screenshots are looking for Usage: sg scan, whereas on my system I see Usage: ast-grep scan.

yes, this is caused by a change to reduce binary size so now the `sg` command is an alias to ast-grep. so the error message mismatches.


Let me know if the color code works or you have new request!

MagicDuck commented:
@HerringtonDarkholme That worked perfectly! Thank you so much, this is amazing support!  ❤️ 

@jayeheffernan FYI, if you want to get rid of the color codes in your day-to-day, cloning the ast-grep repo and (assuming you have rust stable) installing with:
```
cargo install --path crates/cli  
``` 
fixed it for me (until we have an official release).

<img width="836" alt="image" src="https://github.com/user-attachments/assets/e233802a-0e1f-4264-99af-e7a749b370b1" />

jayeheffernan commented:
Thanks @MagicDuck, it's been a pleasure working with you too 🙌  New reddit user, but u/jayeheffernan if you'd like to tag 🙏 Excited to get this out there! 🚀

MagicDuck commented:
@jayeheffernan @HerringtonDarkholme FYI reddit post: https://www.reddit.com/r/neovim/comments/1j65l0f/grugfarnvim_now_has_astgrep_yaml_rules_support/

HerringtonDarkholme commented:
ast-grep 0.36.0 dropped and it can be used with the yaml rule feature now!
